packages/protocol/contracts/L1/libs/LibClaiming.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.18;
+
+import {LibUtils} from "./LibUtils.sol";
+import {TaikoData} from "../../L1/TaikoData.sol";
+import {LibAddress} from "../../libs/LibAddress.sol";
+
+library LibClaiming {
+    using LibAddress for address;
+
+    event ClaimBlockBid(
+        uint256 indexed id,
+        address claimer,
+        uint256 claimedAt,
+        uint256 deposit
+    );
+
+    event BidRefunded(
+        uint256 indexed id,
+        address claimer,
+        uint256 refundedAt,
+        uint256 refund
+    );
+
+    error L1_ID();
+    error L1_CLAIM_AUCTION_WINDOW_PASSED();
+    error L1_HALTED();
+    error L1_ALREADY_CLAIMED();
+    error L1_INVALID_CLAIM_DEPOSIT();
+    error L1_CLAIM_AUCTION_DELAY_PASSED();
+
+    // claimBlock lets a prover claim a block for only themselves to prove,
+    // for a limited amount of time.
+    // if the time passes and they dont submit a proof,
+    // they will lose their deposit, half to the new prover, half is burnt.
+    // there is a set window anyone can bid, then the winner has N time to prove it.
+    // after which anyone can prove.
+    function claimBlock(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        uint256 blockId
+    ) internal {
+        if (
+            state.proposedBlocks[blockId % config.maxNumBlocks].proposer ==
+            address(0)
+        ) {
+            revert L1_ID();
+        }
+
+        TaikoData.Claim memory currentClaim = state.claims[blockId];
+
+        if (
+            currentClaim.claimer != address(0) &&
+            (block.timestamp - currentClaim.claimedAt >
+                config.claimAuctionDelayInSeconds)
+        ) {
+            revert L1_CLAIM_AUCTION_DELAY_PASSED();
+        }
+
+        // we need to make sure this bid
+        // is within it claimAuctionWindowInSeconds.
+        if (
+            block.timestamp -
+                state.proposedBlocks[blockId % config.maxNumBlocks].proposedAt >
+            config.claimAuctionWindowInSeconds
+        ) {
+            revert L1_CLAIM_AUCTION_WINDOW_PASSED();
+        }
+
+        // if user has claimed and not proven a block before, we multiply
+        // requiredDeposit by the number of times they have falsely claimed.
+        uint256 baseDepositRequired = config.baseClaimDepositInWei;
+        if (state.timesProofNotDeliveredForClaim[tx.origin] > 0) {
```

Comment:
shall we use `msg.sender` instead of `tx.origin` here? since other part of our protocol allows proposer / prover to be `msg.sender`?

Reply:
done